### ISSUE TITLE
Race condition: Firewall DNS resolution can take place before the network is up

### ISSUE THREAD

**Issue by strugee** [QubesOS/qubes-issues#5576]
**Qubes OS version**

R4.0

**Affected component(s) or functionality**

Qubes Firewall

**Brief summary**

**To Reproduce**

1. Set up some qube to have a firewall configuration based on hostnames
2. Restart the computer, possibly a couple times
3. Notice how that qube's network access is blocked by the firewall because the firewall tried to query the DNS before the network was up, and it failed closed

**Expected behavior**

Qubes Firewall waits until its network access has properly come up to make those DNS queries. Note that "connected to the NetVM" probably isn't good enough; what we care about is whether the _NetVM_ is able to actually provide access to the internet to the FirewallVM. Note that basically all my connections are via WiFi and association can take a little bit, which I'm sure exacerbates this issue.

Alternately, Qubes Firewall could requery the DNS any time the network connection in the NetVM changed (which would include the initial connection after system boot).

**Actual behavior**

All the DNS queries failed because the network wasn't up and the firewall failed closed and blocked all network access.

**Solutions you've tried**

Opening the settings pane of the relevant qube and then clicking OK causes the firewall rules to be reloaded which in turns triggers DNS queries to be made again. These queries then succeed because the network is up by this time. This functions as a workaround.

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

https://www.qubes-os.org/doc/firewall/

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

#5269

**Comment by Rowofo** [comment]
I'm having similar issue with version R4.0.

#### Issue
Internet connection is blocked using clearnet to all sys-firewall appVMs but sys-whonix appVMs is good.

#### My network setup:
sys-net -->  sys-firewall --> devVM
No internet to all clearnet appVMs connection, but ssh from devVM to local network outside of Qubes PC is working fine.

sys-net -->  sys-firewall --> sys-whonix --> anon-whonix
Working good.

sys-net -->  devVM
I tried to switch directly using No way, the issue continued

@strugee solutions does not help in my case.

System update using sys-firewall on dom0 <code>sudo qubes-dom0-update</code> failed but other templates was able to run updates without issue.  

#### sys-net debugging results:
<code>ping -c 3 google.com</code>
```
PING google.com (172.217.16.206) 56(84) bytes of data.
64 bytes from fra16s08-in-f206.1e100.net (172.217.16.206): icmp_seq=1 ttl=47 time=41.3 ms
64 bytes from fra16s08-in-f206.1e100.net (172.217.16.206): icmp_seq=2 ttl=47 time=38.4 ms
64 bytes from fra16s08-in-f206.1e100.net (172.217.16.206): icmp_seq=3 ttl=47 time=40.7 ms

--- google.com ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 178ms
rtt min/avg/max/mdev = 38.369/40.154/41.346/1.285 ms
```
<code>ping -c 3 duckduckgo.com</code>
```
PING duckduckgo.com (79.125.105.113) 56(84) bytes of data.

--- duckduckgo.com ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 51ms
```
Note: other domains does respond except duckduckgo.com

<code>iptables-save</code>
```
# Generated by iptables-save v1.8.2 on Sat Jan 25 10:25:17 2020
*raw
:PREROUTING ACCEPT [38931:18904286]
:OUTPUT ACCEPT [311:28384]
-A PREROUTING -s 10.137.0.6/32 ! -i vif+ -j DROP
-A PREROUTING ! -s 10.137.0.6/32 -i vif5.0 -j DROP
COMMIT
# Completed on Sat Jan 25 10:25:17 2020
# Generated by iptables-save v1.8.2 on Sat Jan 25 10:25:17 2020
*nat
:PREROUTING ACCEPT [49:16177]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [2:381]
:POSTROUTING ACCEPT [0:0]
:PR-QBS - [0:0]
:PR-QBS-SERVICES - [0:0]
-A PREROUTING -j PR-QBS
-A PREROUTING -j PR-QBS-SERVICES
-A POSTROUTING -o vif+ -j ACCEPT
-A POSTROUTING -o lo -j ACCEPT
-A POSTROUTING -j MASQUERADE
-A PR-QBS -d 10.139.1.1/32 -p udp -m udp --dport 53 -j DNAT --to-destination 172.104.136.243
-A PR-QBS -d 10.139.1.1/32 -p tcp -m tcp --dport 53 -j DNAT --to-destination 172.104.136.243
-A PR-QBS -d 10.139.1.2/32 -p udp -m udp --dport 53 -j DNAT --to-destination 159.89.249.249
-A PR-QBS -d 10.139.1.2/32 -p tcp -m tcp --dport 53 -j DNAT --to-destination 159.89.249.249
-A PR-QBS-SERVICES -d 10.137.255.254/32 -i vif+ -p tcp -m tcp --dport 8082 -j REDIRECT
COMMIT
# Completed on Sat Jan 25 10:25:17 2020
# Generated by iptables-save v1.8.2 on Sat Jan 25 10:25:17 2020
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [2:381]
:QBS-FORWARD - [0:0]
-A INPUT -i vif+ -p tcp -m tcp --dport 8082 -j ACCEPT
-A INPUT -i vif+ -p udp -m udp --dport 68 -j DROP
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i vif+ -p icmp -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -i vif+ -j REJECT --reject-with icmp-host-prohibited
-A INPUT -j DROP
-A FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -j QBS-FORWARD
-A FORWARD -i vif+ -o vif+ -j DROP
-A FORWARD -i vif+ -j ACCEPT
-A FORWARD -j DROP
COMMIT
# Completed on Sat Jan 25 10:25:17 2020
```
<code>iptables -L -v</code>
```
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     tcp  --  vif+   any     anywhere             anywhere             tcp dpt:us-cli
    0     0 DROP       udp  --  vif+   any     anywhere             anywhere             udp dpt:bootpc
  152 17476 ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     icmp --  vif+   any     anywhere             anywhere            
  120  8040 ACCEPT     all  --  lo     any     anywhere             anywhere            
    0     0 REJECT     all  --  vif+   any     anywhere             anywhere             reject-with icmp-host-prohibited
  617  228K DROP       all  --  any    any     anywhere             anywhere            

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
29388   18M ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED
 8527  564K QBS-FORWARD  all  --  any    any     anywhere             anywhere            
    0     0 DROP       all  --  vif+   vif+    anywhere             anywhere            
 8527  564K ACCEPT     all  --  vif+   any     anywhere             anywhere            
    0     0 DROP       all  --  any    any     anywhere             anywhere            

Chain OUTPUT (policy ACCEPT 2 packets, 381 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain QBS-FORWARD (1 references)
 pkts bytes target     prot opt in     out     source               destination         
```


#### sys-firewall debugging results:
<code>ping -c 3 google.com</code>
```
ping: google.com: Name or service not known
```

<code>ping -c 3 duckduckgo.com</code>
```
ping: duckduckgo.com: Name or service not known
```

<code>iptables-save</code>
```
# Generated by iptables-save v1.8.2 on Sat Jan 25 10:58:40 2020
*raw
:PREROUTING ACCEPT [41119:19854204]
:OUTPUT ACCEPT [0:0]
-A PREROUTING -s 10.137.0.8/32 ! -i vif+ -j DROP
-A PREROUTING ! -s 10.137.0.8/32 -i vif8.0 -j DROP
-A PREROUTING -s 10.137.0.9/32 ! -i vif+ -j DROP
-A PREROUTING ! -s 10.137.0.9/32 -i vif7.0 -j DROP
-A PREROUTING -s 10.137.0.21/32 ! -i vif+ -j DROP
-A PREROUTING ! -s 10.137.0.21/32 -i vif6.0 -j DROP
COMMIT
# Completed on Sat Jan 25 10:58:40 2020
# Generated by iptables-save v1.8.2 on Sat Jan 25 10:58:40 2020
*nat
:PREROUTING ACCEPT [53:2862]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:PR-QBS - [0:0]
:PR-QBS-SERVICES - [0:0]
-A PREROUTING -j PR-QBS
-A PREROUTING -j PR-QBS-SERVICES
-A POSTROUTING -o vif+ -j ACCEPT
-A POSTROUTING -o lo -j ACCEPT
-A POSTROUTING -j MASQUERADE
-A PR-QBS -d 10.139.1.1/32 -p udp -m udp --dport 53 -j DNAT --to-destination 10.139.1.1
-A PR-QBS -d 10.139.1.1/32 -p tcp -m tcp --dport 53 -j DNAT --to-destination 10.139.1.1
-A PR-QBS -d 10.139.1.2/32 -p udp -m udp --dport 53 -j DNAT --to-destination 10.139.1.2
-A PR-QBS -d 10.139.1.2/32 -p tcp -m tcp --dport 53 -j DNAT --to-destination 10.139.1.2
COMMIT
# Completed on Sat Jan 25 10:58:40 2020
# Generated by iptables-save v1.8.2 on Sat Jan 25 10:58:40 2020
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [0:0]
:QBS-FORWARD - [0:0]
-A INPUT -i vif+ -p udp -m udp --dport 68 -j DROP
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i vif+ -p icmp -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -i vif+ -j REJECT --reject-with icmp-host-prohibited
-A INPUT -j DROP
-A FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -j QBS-FORWARD
-A FORWARD -i vif+ -o vif+ -j DROP
-A FORWARD -i vif+ -j ACCEPT
-A FORWARD -j DROP
COMMIT
# Completed on Sat Jan 25 10:58:40 2020
```

<code>iptables -L -v</code>
```
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DROP       udp  --  vif+   any     anywhere             anywhere             udp dpt:bootpc
    0     0 ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     icmp --  vif+   any     anywhere             anywhere            
    0     0 ACCEPT     all  --  lo     any     anywhere             anywhere            
    0     0 REJECT     all  --  vif+   any     anywhere             anywhere             reject-with icmp-host-prohibited
    0     0 DROP       all  --  any    any     anywhere             anywhere            

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
31938   19M ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED
 9189  607K QBS-FORWARD  all  --  any    any     anywhere             anywhere            
    0     0 DROP       all  --  vif+   vif+    anywhere             anywhere            
 9189  607K ACCEPT     all  --  vif+   any     anywhere             anywhere            
    0     0 DROP       all  --  any    any     anywhere             anywhere            

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain QBS-FORWARD (1 references)
 pkts bytes target     prot opt in     out     source               destination
```

I have already reboot my system multiple times and I'm still blocked.

**Comment by Rowofo** [comment]
Update

I found out the issue was on <code>sys-net</code> which base on Fedora-30-minimal template, everything started working again after I created new <code>sys-net</code> from Debian-10-minimal.

All appVMs connected to clearnet internet without problem through the same <code>sys-firewall</code>.

**Comment by 3hhh** [comment]
Possibly related: #5570

**Comment by strugee** [comment]
Keyword: race condition (because I used that as a search term to find this issue in the tracker and it didn't come up)

**Comment by icequbes1** [comment]
I have also experienced this - if sys-firewall cannot resolve hostnames because its netvm (sys-net) doesn't have internet by the time `qubes-firewall` executes in sys-firewall, connected qubes with hostname-based rules will have all internet blocked.

This almost leans towards suggesting a "network connectivity" service such as the `https://www.google.com/generate_204` but I would presume most Qubes OS users would be against any sort of connection probing.

What would be more beneficial is a visual indicator for sys-firewall to communicate to the user that "all network traffic has been blocked", or simply the state of the firewall. Perhaps an icon in the tray for each `qubes-firewall`-enabled netvm that shows:

- Firewall is active (is `qubes-firewall` even executing? green vs red)
- Each attached qube's firewall state ('allow all' [green] vs 'limited' [yellow] vs 'all traffic blocked for some reason' [red])

**Comment by DemiMarie** [comment]
Personally, I would be fine with connection probing, provided that:

- The default service respected user’s privacy.
- It is off by default.
- Users can specify a custom URL, either via the command line or the GUI.

**Comment by 3hhh** [comment]
> Personally, I would be fine with connection probing, provided that:
> 
> * The default service respected user’s privacy.
> * It is off by default.
> * Users can specify a custom URL, either via the command line or the GUI.

I agree. However ideally it would be a list of URLs that is picked randomly from. Also DNS connection probing should be less centralistic than HTTP connection probing as it depends on your upstream DNS server.

An example that I used in the past:
https://github.com/3hhh/blib/blob/988451f205b0802fa34a1cd862aa2e447c155306/lib/net#L55

In any case there might not be any network anyway, i.e. there should be some feedback to the user on failure as indicated by @icequbes1.

**Comment by iamahuman** [comment]
Possible duplicate of #3641
